File: Data Structures/Queue/Circular_Queue_Program.cpp
// Write a program to implement Circular Queue.

#include<iostream>
using namespace std;

class Queue{
    public:
        int front, rear, *Arr, cap, size;
        Queue(int capacity){
            cap = capacity;
            Arr = new int[capacity];
            front = rear = -1;
            size = 0;
        }

        bool isEmpty(){
            if (size == 0){
                return true;
            }
            else{
                return false;
            }
        }

        bool isFull(){
            if (size == cap){
                return true;
            }
            else{
                return false;
            }
        }

        void Enqueue(int value){
            if (isFull()){
                cout << "Queue is Full." << endl << endl;
            }
            else{
                rear = (rear + 1) % cap;
                Arr[rear] = value;
                size++;
                if (front == -1){
                    front = rear;
                }
            }
        }

        void Dequeue(){
            if (isEmpty()){
                cout << "Queue is Empty." << endl;
                cout << endl;
            }
            else{
                cout << "Dequeued element is: " << Arr[front] << endl;
                front = (front + 1) % cap;
                size--;
                if (size == 0){
                    front = rear = -1;
                }
            }
        }

        void Peek(){
            if (size == 0){
                cout << "Queue is Empty" << endl << endl;
            }
            else{
                cout << "The front element is: " << Arr[front] << endl << endl;
            }
        }

        void Print(){
            if (size == 0){
                cout << "Queue is Empty" << endl << endl;
            }
            else{
                    
                for (int i = 0; i < size; i++){
                    cout << Arr[(front + 1) % cap] << " ";
                }
                cout << endl;
            }
        }

        ~Queue(){
            delete[] Arr;
        }
};
int main(){
    int capacity, opr;
    cout << "Enter the capacity of the circular queue: ";
    cin >> capacity;
    cout << endl;

    Queue obj(capacity);

    cout << "Enter the number of operations: ";
    cin >> opr;
    cout << endl;

    while (opr--){
        int choice, value;
        cout << "Enter: \n\n1. Enqueue\n2. Dequeue\n3. Peek\n4. Print\n\n";
        cin >> choice;
        switch(choice){
            case 1:
                cout << endl;
                cout << "Enter the value to enqueue: ";
                cin >> value;
                cout << value << " has been enqueued." << endl << endl; 
                obj.Enqueue(value);
                break;
            case 2:
                obj.Dequeue();
                break;
            case 3:
                obj.Peek();
                break;
            case 4:
                obj.Print();
                break;
            default:
                cout << "Invalid choice. Please try again." << endl;
                opr++;
                break;
        }
    
    }

    return 0;
}

File: Data Structures/Queue/Java/Circular_Queue_Program.java
// Write a program to implement Circular Queue.

import java.util.Scanner;

class Queue{
    int front, rear, size, cap;
    int[] queue;
    Queue(int capacity){
        cap = capacity;
        front = rear = -1;
        size = 0;
        queue = new int[cap];
    }
    public boolean isEmpty(){
        if (size == 0)
            return true;
        else
            return false;
    }
    public boolean isFull(){
        if (size == cap)
            return true;
        else
            return false;
    }
    public void Enqueue(int value){
        if (isFull())
            System.out.println("Circular Queue is Full!");
        else{
            rear = (rear + 1) % cap;
            queue[rear] = value;
            size++;
            if (front == -1)
                front = rear;
            System.out.println(value + " has been Enqueued to the Circular Queue.");
        }
    }
    public int Dequeue(){
        if (isEmpty())
            return -1;
        else{
            int value = queue[front];
            front = (front + 1) % cap;
            size--;
            if (size == 0)
                front = rear = -1;
            return value;
        }
    }
    public int Peek(){
        if (isEmpty())
            return -1;
        else
            return queue[front];
    }
    public void print(){
        if (isEmpty())
            System.out.println("Circular Queue is Empty!");
        else{
            System.out.print("Circular Queue elements: ");
            int i = front;
            while (i!= rear){
                System.out.print(queue[i] + " ");
                i = (i + 1) % cap;
            }
            System.out.println(queue[rear]);
        }
    }
}
public class Circular_Queue_Program{
    public static void main(String[] args){
        int capacity, value, choice, opr;
        boolean flag = true;
        Scanner read = new Scanner(System.in);
        System.out.print("Enter the capacity of the Circular Queue: ");
        capacity = read.nextInt();
        Queue obj = new Queue(capacity);
        System.out.print("Enter the number of operations: ");
        opr = read.nextInt();
        while (opr-- > 0){
            System.out.print("\nEnter the operation:\n\n1. Enqueue\n2. Dequeue\n3. Peek\n4. print\n\n");
            choice = read.nextInt();
            switch(choice){
                case 1:
                    System.out.print("Enter the element you want to add in the Circular Queue: ");
                    value = read.nextInt();
                    obj.Enqueue(value);
                    break;
                case 2: 
                    flag = obj.isEmpty();
                    value = obj.Dequeue();
                    if (value == -1 && flag == true)
                        System.out.println("Circular Queue is Empty!");
                    else
                        System.out.println(value + " has been Dequeued from the Circular Queue.");
                    break;
                case 3:
                    flag = obj.isEmpty();
                    value = obj.Peek();
                    if (value == -1 && flag == true)
                        System.out.println("Circular Queue is Empty!");
                    else
                        System.out.println(value + " is the front element of the Circular Queue.");
                    break;
                case 4:
                    obj.print();
                    break;
                default:
                    System.out.println("Invalid Operation!");
                    opr++;
                    break;
            }
        }
    }
}